cpp/src/arrow/type.cc

```diff
@@ -2090,17 +2090,39 @@ Status SchemaBuilder::AreCompatible(const std::vector<std::shared_ptr<Schema>>&
   return Merge(schemas, policy).status();
 }
 
+std::vector<std::shared_ptr<Field>> make_fields(
+    std::initializer_list<std::pair<std::string, std::shared_ptr<DataType>>> init_list) {
+  std::vector<std::shared_ptr<Field>> fields;
+  fields.reserve(init_list.size());
+  for (const auto& [name, type] : init_list) {
+    fields.push_back(field(name, type));
+  }
```

Comment:
Not sure this will work, but worth trying:
```suggestion
  for (auto&& [name, type] : init_list) {
    fields.push_back(field(std::move(name), std::move(type)));
  }
```

@bkietz Thoughts?

Reply:
I will take a look which one is better using a similar demo at compiler explorer tomorrow.

Reply:
The elements of an initializer list are always const, so calling move() on them won't accomplish anything. We could elide the copy by making the initializer lists accept string_view instead of string, though: 
```
std::vector<std::shared_ptr<Field>> MakeFields(
    std::initializer_list<std::pair<std::string_view, std::shared_ptr<DataType>>> init_list) {
  for (const auto& [name, type] : init_list) {
    fields.push_back(field(std::string{name}, type));
  }
}
```

However I've seldom seen a field name longer than 22 chars, so they'd be inlined in most STL impls.

From a perf golf standpoint, the larger annoyance is the type, which must be copied (since we don't have a nice implicitly constructible non-owning DataType holder) and which hits atomics when copied. If you wanted to get around *that* you'd need something like:
```
struct DataTypeView {
  // NOLINTNEXTLINE
  DataTypeView(const std::shared_ptr<DataType>& type) : type{type} {}
  // NOLINTNEXTLINE
  operator std::shared_ptr<DataType>() const { return type; }
  const std::shared_ptr<DataType>& type;
};
std::vector<std::shared_ptr<Field>> MakeFields(
    std::initializer_list<std::pair<std::string_view, DataTypeView>> init_list) {
  for (const auto& [name, type] : init_list) {
    fields.push_back(field(std::string{name}, type));
  }
}
```

Reply:
I see, thank you @bkietz !
Well, let's not overdo this and keep the const-ref instead.